File: shed/game/card.py
from rlcard.games.base import Card


class ShedCard(Card):
    valid_suit = ["S", "H", "D", "C", "BJ", "RJ"]
    valid_rank = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K"]
    rank_to_value = {
        "2": 2,
        "4": 4,
        "5": 5,
        "6": 6,
        "7": 7,
        "8": 8,
        "9": 9,
        "J": 11,
        "Q": 12,
        "K": 13,
        "T": 14,
        "3": 15,
        "A": 16,
    }

    def __init__(self, suit, rank):
        super().__init__(suit, rank)

    def is_magic_card(self) -> bool:
        return self.rank in ["A", "3", "T", "7"]

    def is_ace(self) -> bool:
        return self.rank == "A"

    def is_ten(self) -> bool:
        return self.rank == "T"

    def is_seven(self) -> bool:
        return self.rank == "7"

    def is_three(self) -> bool:
        return self.rank == "3"

    def __eq__(self, other: "ShedCard") -> bool:
        if isinstance(other, Card):
            return self.rank == other.rank
        else:
            raise ValueError("Can only compare to other ShedCard")

    def __lt__(self, other: "ShedCard") -> bool:
        if other.is_ace() and not self.is_magic_card():
            return False

        if isinstance(other, Card):
            value_self = self.rank_to_value[self.rank]
            value_other = self.rank_to_value[other.rank]
            return value_self < value_other
        else:
            raise ValueError("Can only compare to other ShedCard")

    def __le__(self, other: "ShedCard") -> bool:
        return self.__lt__(other) or self.__eq__(other)

    def __gt__(self, other: "ShedCard") -> bool:
        if self.is_ace():
            return True

        if isinstance(other, Card):
            value_self = self.rank_to_value[self.rank]
            value_other = self.rank_to_value[other.rank]
            return value_self > value_other
        else:
            raise ValueError("Can only compare to other ShedCard")

    def __ge__(self, other: "ShedCard") -> bool:
        return self.__gt__(other) or self.__eq__(other)

File: rust/shed/src/game/card.rs
use std::{
    cmp::Ordering,
    fmt::{self},
};
use strum_macros::EnumIter;

use super::action::Action;

#[derive(Eq, PartialEq, Debug, EnumIter, Copy, Clone, Hash)]
pub enum Suit {
    Clubs,
    Diamonds,
    Hearts,
    Spades,
}

#[derive(Eq, PartialEq, Debug, EnumIter, Copy, Clone, Hash)]
pub enum Rank {
    Ace,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Ten,
    Jack,
    Queen,
    King,
}

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Suit::Clubs => write!(f, "C"),
            Suit::Diamonds => write!(f, "D"),
            Suit::Hearts => write!(f, "H"),
            Suit::Spades => write!(f, "S"),
        }
    }
}

impl fmt::Display for Rank {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Rank::Ace => write!(f, "A"),
            Rank::Two => write!(f, "2"),
            Rank::Three => write!(f, "3"),
            Rank::Four => write!(f, "4"),
            Rank::Five => write!(f, "5"),
            Rank::Six => write!(f, "6"),
            Rank::Seven => write!(f, "7"),
            Rank::Eight => write!(f, "8"),
            Rank::Nine => write!(f, "9"),
            Rank::Ten => write!(f, "T"),
            Rank::Jack => write!(f, "J"),
            Rank::Queen => write!(f, "Q"),
            Rank::King => write!(f, "K"),
        }
    }
}

#[derive(Eq, Debug, Hash, Clone)]
pub struct Card {
    suit: Suit,
    rank: Rank,
}

impl Card {
    pub fn new(suit: Suit, rank: Rank) -> Self {
        Self { suit, rank }
    }

    pub fn suit(&self) -> &Suit {
        &self.suit
    }

    pub fn rank(&self) -> &Rank {
        &self.rank
    }

    pub fn get_index(&self) -> String {
        format!("{}{}", self.suit, self.rank)
    }

    pub fn is_ace(&self) -> bool {
        self.rank == Rank::Ace
    }

    pub fn is_ten(&self) -> bool {
        self.rank == Rank::Ten
    }

    pub fn is_seven(&self) -> bool {
        self.rank == Rank::Seven
    }

    pub fn is_three(&self) -> bool {
        self.rank == Rank::Three
    }

    pub fn is_magic_card(&self) -> bool {
        self.is_ace() || self.is_three() || self.is_seven() || self.is_ten()
    }

    pub fn value(&self) -> u16 {
        match &self.rank {
            Rank::Two => 2,
            Rank::Three => 3,
            Rank::Four => 4,
            Rank::Five => 5,
            Rank::Six => 6,
            Rank::Seven => 7,
            Rank::Eight => 8,
            Rank::Nine => 9,
            Rank::Ten => 10,
            Rank::Jack => 11,
            Rank::Queen => 12,
            Rank::King => 13,
            Rank::Ace => 14,
        }
    }

    pub fn to_action(&self) -> Action {
        match &self.rank {
            Rank::Ace => Action::Ace,
            Rank::Two => Action::Two,
            Rank::Three => Action::Three,
            Rank::Four => Action::Four,
            Rank::Five => Action::Five,
            Rank::Six => Action::Six,
            Rank::Seven => Action::Seven,
            Rank::Eight => Action::Eight,
            Rank::Nine => Action::Nine,
            Rank::Ten => Action::Ten,
            Rank::Jack => Action::Jack,
            Rank::Queen => Action::Queen,
            Rank::King => Action::King,
        }
    }

}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.get_index())
    }
}

impl PartialOrd for Card {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self.lt(other) {
            return Some(Ordering::Less);
        };
        if self.le(other) {
            return Some(Ordering::Equal);
        };
        if self.ge(other) {
            return Some(Ordering::Equal);
        };
        if self.gt(other) {
            return Some(Ordering::Greater);
        };
        None
    }

    fn lt(&self, other: &Self) -> bool {
        if other.is_ace() && !self.is_magic_card() {
            return true;
        };
        self.value().lt(&other.value())
    }

    fn le(&self, other: &Self) -> bool {
        self.eq(other) || self.lt(other)
    }

    fn gt(&self, other: &Self) -> bool {
        if self.is_ace() {
            return true;
        };
        self.value().gt(&other.value())
    }

    fn ge(&self, other: &Self) -> bool {
        self.eq(other) || self.gt(other)
    }
}

impl PartialEq for Card {
    fn eq(&self, other: &Self) -> bool {
        self.rank == other.rank
    }
}
